test/model/src/core/kv_table_item.cpp

```diff
@@ -40,18 +42,77 @@ namespace model {
  */
 int
 kv_table_item::add_update(kv_update &&update, bool must_exist, bool must_not_exist)
+{
+    std::shared_ptr<kv_update> update_ptr = std::make_shared<kv_update>(std::move(update));
+    return add_update(update_ptr, must_exist, must_not_exist);
+}
+
+/*
+ * kv_table_item::add_update --
+ *     Add an update.
+ */
+int
+kv_table_item::add_update(std::shared_ptr<kv_update> update, bool must_exist, bool must_not_exist)
 {
     std::lock_guard lock_guard(_lock);
+    int ret = add_update_nolock(update, must_exist, must_not_exist);
+    if (ret == WT_ROLLBACK)
+        update->txn()->fail();
+    return ret;
+}
+
+/*
+ * kv_table_item::add_update_nolock --
+ *     Add an update but without taking a lock (this assumes the caller has it).
+ */
+int
+kv_table_item::add_update_nolock(
+  std::shared_ptr<kv_update> update, bool must_exist, bool must_not_exist)
+{
     kv_update::timestamp_comparator cmp;
 
     /* If this is a non-timestamped update, there cannot be existing timestamped updates. */
-    if (update.global()) {
+    if (update->global()) {
         if (!_updates.empty() && !_updates.back()->global())
             return EINVAL;
     }
 
+    /* Check for transaction conflicts. */
+    kv_transaction_ptr txn = update->txn();
+    if (txn) {
+        /*
+         * Check whether the update chain has an update that is not included in this transaction's
+         * snapshot, or if it has an uncommitted update (regardless of whether it is included in its
+         * transaction snapshot.
+         */
+        for (auto &u : _updates) {
+            switch (u->txn_state()) {
+            case kv_transaction_state::in_progress:
+                /* Can't conflict with self. */
+                if (u->txn_id() == txn->id()) {
+                    /*
+                     * Cannot update a key with a more recent timestamp. This is where the model
+                     * diverges from WiredTiger: We return an error here, but WiredTiger would abort
+                     * during commit.
+                     */
```

Comment:
I changed this to throwing `model::wiredtiger_abort_exception`, which is an exception that I just introduced to model that the same behavior in WiredTiger would result in abort. And so that we can test this, I introduced `model_testutil_assert_exception` to check for that behavior (defined in `model/test/util.h`.

The original intention was to not abort so that we can actually test this—but throwing an exception would in this case make much more sense.